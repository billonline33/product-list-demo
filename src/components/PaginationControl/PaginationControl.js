import React from 'react';
import { Link } from 'react-router-dom';
import { getProductPageUrl } from '../../utils';

const PaginationControl = props => {
  const { pageNumber, pageSize, totalProducts, maxPageNumbersToShow } = props;
  const pageNumbers = [];
  const nPages = maxPageNumbersToShow;
  const lastPage = Math.ceil(totalProducts / pageSize);

  for (var i = pageNumber; i < pageNumber + nPages; i++) {
    if (i <= lastPage) {
      pageNumbers.push(i);
    }
  }

  return (
    <div className="listing_navigation">
      <Link
        to={getProductPageUrl(pageNumber - 1, pageSize)}
        className={pageNumber === 1 ? 'listing_navigation__disabled' : ''}
      >
        {'<'} Previous page{' '}
      </Link>
      {pageNumbers.map(item => {
        return (
          <Link
            to={getProductPageUrl(item, pageSize)}
            className={pageNumber === item ? 'listing_navigation__active' : ''}
            key={item}
          >
            {item}
          </Link>
        );
      })}
      <Link
        to={getProductPageUrl(pageNumber + 1, pageSize)}
        className={
          pageNumber === lastPage ? 'listing_navigation__disabled' : ''
        }
      >
        Next page >
      </Link>
    </div>
  );
};

export default PaginationControl;
