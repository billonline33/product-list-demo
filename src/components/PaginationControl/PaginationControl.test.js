import React from 'react';
import PaginationControl from './PaginationControl';
import { shallow } from 'enzyme';

describe('PaginationControl component', () => {
  const mockProps = {
    pageNumber: 1,
    pageSize: 8,
    totalProducts: 16,
    maxPageNumbersToShow: 4
  };

  let wrapper = shallow(<PaginationControl {...mockProps} />);

  it('PaginationControl renders without crashing', () => {
    expect(shallow(<PaginationControl {...mockProps} />)).toMatchSnapshot();
  });

  it('PaginationControl first link has class listing_navigation__disabled if pageNumber = 1', () => {
    expect(wrapper.props().children[0].props.className).toEqual(
      'listing_navigation__disabled'
    );
  });

  const mockProps2 = {
    pageNumber: 8,
    pageSize: 8,
    totalProducts: 64,
    maxPageNumbersToShow: 4
  };

  const wrapper2 = shallow(<PaginationControl {...mockProps2} />);
  it('PaginationControl last link has class listing_navigation__disabled if pageNumber = lastPage', () => {
    expect(wrapper2.props().children[0].props.className).toEqual('');
    expect(wrapper2.props().children[2].props.className).toEqual(
      'listing_navigation__disabled'
    );
  });
});
