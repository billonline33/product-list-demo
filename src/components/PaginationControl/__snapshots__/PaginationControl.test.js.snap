// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationControl component PaginationControl renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PaginationControl
    maxPageNumbersToShow={4}
    pageNumber={1}
    pageSize={8}
    totalProducts={16}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Link
          className="listing_navigation__disabled"
          replace={false}
          to="/?page=0&per_page=8"
>
          &lt;
           Previous page
           
</Link>,
        Array [
          <Link
            className="listing_navigation__active"
            replace={false}
            to="/?page=1&per_page=8"
>
            1
</Link>,
          <Link
            className=""
            replace={false}
            to="/?page=2&per_page=8"
>
            2
</Link>,
        ],
        <Link
          className=""
          replace={false}
          to="/?page=2&per_page=8"
>
          Next page 
          &gt;
</Link>,
      ],
      "className": "listing_navigation",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            "<",
            " Previous page",
            " ",
          ],
          "className": "listing_navigation__disabled",
          "replace": false,
          "to": "/?page=0&per_page=8",
        },
        "ref": null,
        "rendered": Array [
          "<",
          " Previous page",
          " ",
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "children": 1,
          "className": "listing_navigation__active",
          "replace": false,
          "to": "/?page=1&per_page=8",
        },
        "ref": null,
        "rendered": 1,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "children": 2,
          "className": "",
          "replace": false,
          "to": "/?page=2&per_page=8",
        },
        "ref": null,
        "rendered": 2,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            "Next page ",
            ">",
          ],
          "className": "",
          "replace": false,
          "to": "/?page=2&per_page=8",
        },
        "ref": null,
        "rendered": Array [
          "Next page ",
          ">",
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Link
            className="listing_navigation__disabled"
            replace={false}
            to="/?page=0&per_page=8"
>
            &lt;
             Previous page
             
</Link>,
          Array [
            <Link
              className="listing_navigation__active"
              replace={false}
              to="/?page=1&per_page=8"
>
              1
</Link>,
            <Link
              className=""
              replace={false}
              to="/?page=2&per_page=8"
>
              2
</Link>,
          ],
          <Link
            className=""
            replace={false}
            to="/?page=2&per_page=8"
>
            Next page 
            &gt;
</Link>,
        ],
        "className": "listing_navigation",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "<",
              " Previous page",
              " ",
            ],
            "className": "listing_navigation__disabled",
            "replace": false,
            "to": "/?page=0&per_page=8",
          },
          "ref": null,
          "rendered": Array [
            "<",
            " Previous page",
            " ",
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "children": 1,
            "className": "listing_navigation__active",
            "replace": false,
            "to": "/?page=1&per_page=8",
          },
          "ref": null,
          "rendered": 1,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "children": 2,
            "className": "",
            "replace": false,
            "to": "/?page=2&per_page=8",
          },
          "ref": null,
          "rendered": 2,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "Next page ",
              ">",
            ],
            "className": "",
            "replace": false,
            "to": "/?page=2&per_page=8",
          },
          "ref": null,
          "rendered": Array [
            "Next page ",
            ">",
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
